src/wallet/wallet2.cpp

```diff
@@ -11448,29 +11529,48 @@ uint64_t wallet2::import_key_images(const std::vector<std::pair<crypto::key_imag
   return m_transfers[signed_key_images.size() - 1].m_block_height;
 }
 
-bool wallet2::import_key_images(std::vector<crypto::key_image> key_images)
+bool wallet2::import_key_images(std::vector<crypto::key_image> key_images, size_t offset, boost::optional<std::unordered_set<size_t>> selected_transfers)
 {
-  if (key_images.size() > m_transfers.size())
+  if (key_images.size() + offset > m_transfers.size())
   {
     LOG_PRINT_L1("More key images returned that we know outputs for");
     return false;
   }
   for (size_t i = 0; i < key_images.size(); ++i)
   {
-    transfer_details &td = m_transfers[i];
+    const size_t idx = i + offset;
+    if (selected_transfers && selected_transfers.get().find(idx) == selected_transfers.get().end())
+      continue;
+
+    transfer_details &td = m_transfers[idx];
     if (td.m_key_image_known && !td.m_key_image_partial && td.m_key_image != key_images[i])
       LOG_PRINT_L0("WARNING: imported key image differs from previously known key image at index " << i << ": trusting imported one");
     td.m_key_image = key_images[i];
-    m_key_images[m_transfers[i].m_key_image] = i;
+    m_key_images[td.m_key_image] = i;
     td.m_key_image_known = true;
     td.m_key_image_request = false;
     td.m_key_image_partial = false;
-    m_pub_keys[m_transfers[i].get_public_key()] = i;
+    m_pub_keys[td.get_public_key()] = i;
```

Comment:
That seems wrong. Those are supposed index into m_transfers. Are you sure you don't mean to use "= idx" on these two ?

Reply:
The idea is that `key_images` can be shorter than `m_transfers`, specified by the offset, like with `export_outputs`.

```cpp
const size_t idx = i + offset;
transfer_details &td = m_transfers[idx];
```

Reply:
Yes, but here:
m_pub_keys[td.get_public_key()] = i;

That i is going to offset m_transfers, it's supposed to point back to the transfer_details structure. AFAICT, it's going to point to the wrong one here.

Reply:
oh, I missed that. You are right! Thanks!

Reply:
I've changed var names to ki_idx and transfer_idx so it is clear now